### PR TITLE
MiscellaneousChangeForRntbd

### PR DESCRIPTION
**Pull request by xinlian12**
This PR is a duplicate of changes not related to ConnectionStateListener incuded in PR #14697. For example, renaming, style clean etc

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java [hunk 1]
```diff
@@ -151,19 +151,17 @@ public String toString() {
...
 
-    @JsonPropertyOrder({ "name", "startTimeUTC", "durationInMicroSec" })
+    @JsonPropertyOrder({ "name", "startTimeUTC" })
     public static final class Event {
 
         @JsonIgnore
         private final Duration duration;
 
-        @JsonSerialize(using = ToStringSerializer.class)
-        private final long durationInMicroSec;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:161]: "It is reported, just not stored as a value" (copied from the original PR).

This will not change the rntbd timeline diagnostics. After this change:
"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.239","endTimeUTC":"22 Sep 2020 17:40:59.239","durationInMicroSec":0},{"serializationType":"PARTITION_KEY_FETCH_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.317","endTimeUTC":"22 Sep 2020 17:40:59.323","durationInMicroSec":5998}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"60586 KB","availableMemory":"8309590 KB","systemCpuLoad":"(2020-09-22T17:40:58.729927700Z 20.1%)"}}

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java [hunk 2]
```diff
@@ -327,9 +364,6 @@ private void throwIfClosed() {
         @JsonProperty()
         private final Duration receiveHangDetectionTime;
 
-        @JsonProperty()
-        private final Duration requestExpiryInterval;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:331]: This is removed because no where is using it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java [hunk 3]
```diff
@@ -566,7 +599,6 @@ public String toString() {
          *   "maxRequestsPerChannel": 30,
          *   "maxConcurrentRequestsPerEndpointOverride": 500,
          *   "receiveHangDetectionTime": "PT1M5S",
-         *   "requestExpiryInterval": "PT5S",
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:569]: I think we should remove these, since DirectConnectionConfig already provides the APIs now. This was before the public APIs and was only used through JVM options. 
Or if we want to keep these configurations, we should update them to the correct values - like `maxChannelsPerEndpoint`, `maxRequestsPerChannel`, `idleEndpointTimeout`, etc.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 4]
```diff
@@ -527,7 +527,7 @@ public void write(final ChannelHandlerContext context, final Object message, fin
 
         if (message == RntbdHealthCheckRequest.MESSAGE) {
 
-            context.write(RntbdHealthCheckRequest.MESSAGE, promise).addListener(completed -> {
+            context.writeAndFlush(RntbdHealthCheckRequest.MESSAGE, promise).addListener(completed -> {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:530]: What is the reason behind this change ?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:530]: It is not a required change. But generally write is only queue the message but not really send through wire, writeAndFlush will help flush the message further down the pipeline, so I think it makes sense to keep this change. 

But I am also open to exclude this change for now and may have a PR in future separately for this

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:530]: In any case, there is one more `context.write` we should make sure to flush that one as well. 

If we want to change it, do you know when were we doing the flush in just `write` case ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java [hunk 5]
```diff
@@ -114,9 +113,14 @@ protected void initChannel(final Channel channel) {
...
-            // Initialize sslHandler with jdkCompatibilityMode = true for openssl context.
-            new SslHandler(this.config.sslContext().newEngine(channel.alloc())),
+            // Initialize sslHandler with jdkCompatibilityMode = true for OpenSSL context
+            // TODO Log an issue with netty for clarification on the design of this constructor and the
+            //  semantic differences between the JDK and OpenSSL implementations
+            new SslHandler(this.config.sslContext().newEngine(channel.alloc())));
+
+        final long idleConnectionTimerResolutionInNanos = config.idleConnectionTimerResolutionInNanos();
+
+        pipeline.addFirst(
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java:123]: this adds IdleStateHandler before SSLHandler. Is is intentional?

The old code had IdleStateHandler after SSLHandler.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 4 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:530]: This potentially can be a perf impacting change.
If we decide to include this we should study the impact of this on both latency and throughput. I think the following studies would be needed:
1. both latency and throughput using the normal benchmark flow
2. both latency and throughput on sparse workload (https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/Configuration.java#L144)

**Review by moderakh — COMMENTED** [review]
write <-> writeAndFlush change can potentially can be a perf impacting change.

If we decide to include this we should study the impact of this on both latency and throughput. I think the following studies would be needed:

1. both latency and throughput using the normal benchmark flow (when saturating max load).
2. both latency and throughput on sparse workload (https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/Configuration.java#L144)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java (hunk 5 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java:123]: I believe SSLHandler should be the very first one in the pipeline, no ?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 4 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:530]: I agree, If we want to change it, we should understand how the flushing was working so far and why we want to change it ?
Because I remember in my past conversations with David, flushing immediately had some downsides as well. I can't remember the exact downsides, may be they are not present any more. But a perf check would help us figure it out.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java]: good catch, I do not thinkthat is the intent here, reverted back.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java]: @kushagraThapar My current understanding is that, when we call write, the message will be queued in the out buffer, but it is not for sure that the message will be written to the socket. Netty will decide when to flush the message out, so that is why we do not have to call flush in our code. 

@moderakh  Yea, make sense, and those are good tests to do for this change.

@kushagraThapar  @moderakh  I think I will revert this change for now, so this PR only contains style/renaming change,  and I will have a separate PR for this change and to investigate more about it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/RntbdTransportClient.java:569]: hmm, I think this comments are more about giving different ways how you can using the JVM options and setup the values. I feel it has been very useful when doing tests. 

So for now, I left the comments to be the way it is today, but update the values to be consistent with the default values.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 4 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java]: Makes sense @xinlian12 , thanks.

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks!

**Review by moderakh — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelHandler.java]: @kushagraThapar  yup, SSLHandler is the very first one in the pipeline today, but after the change, IdleStateHandler will be the first one, so did not think that is the intent. Has reverted the change back.

Thanks~
